File: flutter_app/lib/features/routines/domain/rutina.dart
class EjercicioRutina {
  final String id;
  final String rutinaId;
  final String ejercicioId;
  final int orden;
  final int series;
  final String repeticiones;
  final int? descansoSegundos;
  final String? tempo;
  final String? notas;
  final DateTime createdAt;

  const EjercicioRutina({
    required this.id,
    required this.rutinaId,
    required this.ejercicioId,
    required this.orden,
    required this.series,
    required this.repeticiones,
    this.descansoSegundos,
    this.tempo,
    this.notas,
    required this.createdAt,
  });

  factory EjercicioRutina.fromJson(Map<String, dynamic> json) {
    return EjercicioRutina(
      id: json['id'] as String,
      rutinaId: json['rutina_id'] as String,
      ejercicioId: json['ejercicio_id'] as String,
      orden: json['orden'] as int,
      series: json['series'] as int,
      repeticiones: json['repeticiones'] as String,
      descansoSegundos: json['descanso_segundos'] as int?,
      tempo: json['tempo'] as String?,
      notas: json['notas'] as String?,
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }
}

class Rutina {
  final String id;
  final String nombre;
  final String? descripcion;
  final String tipoRutina;
  final String? dificultad;
  final int? duracionEstimadaMinutos;
  final int frecuenciaSemanal;
  final String? usuarioId;
  final String? creadorId;
  final bool esPublica;
  final String fuenteCreacion;
  final Map<String, dynamic>? metadataIa;
  final DateTime createdAt;
  final DateTime updatedAt;
  final bool activa;
  final List<EjercicioRutina> ejercicios;

  const Rutina({
    required this.id,
    required this.nombre,
    this.descripcion,
    required this.tipoRutina,
    this.dificultad,
    this.duracionEstimadaMinutos,
    this.frecuenciaSemanal = 3,
    this.usuarioId,
    this.creadorId,
    this.esPublica = false,
    this.fuenteCreacion = 'entrenador',
    this.metadataIa,
    required this.createdAt,
    required this.updatedAt,
    this.activa = true,
    this.ejercicios = const [],
  });

  factory Rutina.fromJson(Map<String, dynamic> json) {
    return Rutina(
      id: json['id'] as String,
      nombre: json['nombre'] as String,
      descripcion: json['descripcion'] as String?,
      tipoRutina: json['tipo_rutina'] as String,
      dificultad: json['dificultad'] as String?,
      duracionEstimadaMinutos: json['duracion_estimada_minutos'] as int?,
      frecuenciaSemanal: json['frecuencia_semanal'] as int? ?? 3,
      usuarioId: json['usuario_id'] as String?,
      creadorId: json['creador_id'] as String?,
      esPublica: json['es_publica'] as bool? ?? false,
      fuenteCreacion: json['fuente_creacion'] as String? ?? 'entrenador',
      metadataIa: json['metadata_ia'] as Map<String, dynamic>?,
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: DateTime.parse(json['updated_at'] as String),
      activa: json['activa'] as bool? ?? true,
      ejercicios: (json['ejercicios'] as List<dynamic>?)
              ?.map((e) => EjercicioRutina.fromJson(e as Map<String, dynamic>))
              .toList() ??
          [],
    );
  }

  Map<String, dynamic> toCreateJson() {
    return {
      'nombre': nombre,
      'descripcion': descripcion,
      'tipo_rutina': tipoRutina,
      'dificultad': dificultad,
      'duracion_estimada_minutos': duracionEstimadaMinutos,
      'frecuencia_semanal': frecuenciaSemanal,
      'es_publica': esPublica,
      'fuente_creacion': fuenteCreacion,
      'ejercicios': ejercicios.map((e) => {
        'ejercicio_id': e.ejercicioId,
        'orden': e.orden,
        'series': e.series,
        'repeticiones': e.repeticiones,
        'descanso_segundos': e.descansoSegundos,
        'tempo': e.tempo,
        'notas': e.notas,
      }).toList(),
    };
  }
}

File: backend/app/models/rutina.py
from datetime import datetime

from sqlalchemy import Boolean, DateTime, ForeignKey, Integer, String, Text
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.models.base import Base


class Rutina(Base):
    __tablename__ = "rutinas"

    id: Mapped[str] = mapped_column(
        UUID(as_uuid=False), primary_key=True, default=lambda: str(uuid.uuid4())
    )
    nombre: Mapped[str] = mapped_column(String(150), nullable=False)
    descripcion: Mapped[str | None] = mapped_column(Text, nullable=True)
    tipo_rutina: Mapped[str] = mapped_column(String(30), nullable=False)
    dificultad: Mapped[str | None] = mapped_column(String(20), nullable=True)
    duracion_estimada_minutos: Mapped[int | None] = mapped_column(Integer, nullable=True)
    frecuencia_semanal: Mapped[int] = mapped_column(Integer, default=3)

    usuario_id: Mapped[str | None] = mapped_column(
        UUID(as_uuid=False), ForeignKey("usuarios.id"), nullable=True
    )
    creador_id: Mapped[str | None] = mapped_column(
        UUID(as_uuid=False), ForeignKey("usuarios.id"), nullable=True
    )
    es_publica: Mapped[bool] = mapped_column(Boolean, default=False)

    fuente_creacion: Mapped[str] = mapped_column(String(20), default="entrenador")
    metadata_ia: Mapped[dict | None] = mapped_column(JSONB, nullable=True)

    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)
    updated_at: Mapped[datetime] = mapped_column(
        DateTime, default=datetime.utcnow, onupdate=datetime.utcnow
    )
    activa: Mapped[bool] = mapped_column(Boolean, default=True)

    usuario: Mapped["Usuario | None"] = relationship("Usuario", back_populates="rutinas", foreign_keys=[usuario_id])
    creador: Mapped["Usuario | None"] = relationship("Usuario", foreign_keys=[creador_id])
    ejercicios: Mapped[list["RutinaEjercicio"]] = relationship(
        "RutinaEjercicio", back_populates="rutina", cascade="all, delete-orphan", order_by="RutinaEjercicio.orden"
    )


class RutinaEjercicio(Base):
    __tablename__ = "rutina_ejercicios"

    id: Mapped[str] = mapped_column(
        UUID(as_uuid=False), primary_key=True, default=lambda: str(uuid.uuid4())
    )
    rutina_id: Mapped[str] = mapped_column(
        UUID(as_uuid=False), ForeignKey("rutinas.id", ondelete="CASCADE"), nullable=False
    )
    ejercicio_id: Mapped[str] = mapped_column(
        UUID(as_uuid=False), ForeignKey("ejercicios.id"), nullable=False
    )
    orden: Mapped[int] = mapped_column(Integer, nullable=False, default=1)
    series: Mapped[int] = mapped_column(Integer, nullable=False, default=3)
    repeticiones: Mapped[str] = mapped_column(String(30), nullable=False, default="10-12")
    descanso_segundos: Mapped[int | None] = mapped_column(Integer, nullable=True, default=90)
    tempo: Mapped[str | None] = mapped_column(String(10), nullable=True)
    notas: Mapped[str | None] = mapped_column(Text, nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)

    rutina: Mapped["Rutina"] = relationship("Rutina", back_populates="ejercicios")
    ejercicio: Mapped["Ejercicio"] = relationship("Ejercicio")


import uuid  # noqa: E402, F401
from app.models.usuario import Usuario  # noqa: E402, F401
from app.models.ejercicio import Ejercicio  # noqa: E402, F401
